test/tests/functional/pbs_alps_request_tunables.py

```diff
@@ -0,0 +1,150 @@
+                nline, line_current = data
+                time_current = self.get_epoch(line_current)
+                self.assertGreaterEqual(time_current - time_prev,
+                                        math.ceil(ari),
+                                        msg="alps_release_interval not working"
+                                        )
+                time_prev = time_current
+
+        else:
+            self.skipTest("Reservation cancelled without retry, Try again!")
+
+    def test_alps_release_jitter(self):
+        """
+        Set alps_release_jitter to a higher value and then notice that
+        subsequest reservation cancellation requests are made by adding
+        a random time interval (less than jitter) to alps_release_interval.
+        """
+        arj = 2.198
+        ari = 1
+        max_delay = (ari + math.ceil(2.198))
+        self.mom.add_config({'$alps_release_jitter': arj})
+        self.mom.add_config({'$alps_release_interval': ari})
+        # There is no good way to test jitter and it is a random number
+        # less than value set in alps_release_jitter. So in this case
+        # we can probably try deleting a reservation a few times.
+        for each in range(5):
+            # submit a job and then delete it after it starts running
+            start_time = time.time()
+            j1 = Job(TEST_USER)
+            jid1 = self.server.submit(j1)
+            self.server.expect(JOB, {ATTR_state: 'R'}, id=jid1)
+            self.server.delete(jid1)
+
+            # Look for a message that confirms that reservation is deleted
+            self.mom.log_match("%s;ALPS reservation cancelled" % jid1,
+                               starttime=start_time, max_attempts=5,
+                               interval=2)
+            # Now that we know that reservation is cleared we should
+            # check for time difference between each cancellation request
+            out = self.mom.log_match("%s;Canceling ALPS reservation *" % jid1,
+                                     n='ALL', max_attempts=2, regexp=True,
+                                     allmatch=True)
+
+            # We found something, Let's first check there are atleast 2 such
+            # log messages, If not then that means reservation was cancelled
+            # in the first attempt itself, at that point right thing to do is
+            # to either run it again or find out a way to delay the reservation
+            # cancellation at ALPS level itself.
+            if len(out) >= 2:
+                nline_prev, line_prev = out[0]
+                time_prev = self.get_epoch(line_prev)
+                for data in out[1:]:
+                    nline, line_current = data
+                    time_current = self.get_epoch(line_current)
+                    self.assertLessEqual(time_current - time_prev, max_delay,
+                                         msg="alps_release_jitter not working")
+                    time_prev = time_current
```

Comment:
In this test, you don't skip if you only find one cancellation request...But shouldn't you at least keep a count and if for all 5 jobs you only find one cancellation then you should skip/error?
Which does make me wonder, is skipping the test the right thing to do?  Since we're trying to test a base functionality, wouldn't it be better to error out so someone can look into why the wait_time/jitter code isn't being executed?

Reply:
Well, it gets us into a situation where we can not really say that functionality did not work. I thought for the first case it was probably ok to skip because then we can run it again or at the very least the test will not count as an error. For this test, I was not sure if I should be skipping it because there might be other retries where it might exercise the functionality.
Like you said, I can probably keep a count of retries, and skip if it did not exercise the code in any of the retries. I specifically don't want to error out because it is not PBS's fault if the first cancellation call succeeds.

Reply:
Okay, I see what you mean.

Reply:
The usual reason to skip a test is if the environment you require for the test is not available.  For example, you don't have a Cray or you need 3 moms and only have one.  The problem here is that you're testing a race condition.  If the reservation is always deleted on the first try, then PBS didn't do anything wrong.  We still didn't test what we wanted to test.  This is a hard question to answer.  If we fail, it makes it sound like PBS did something wrong.  If we skip, it makes it sound like we don't have the right environment.  Neither are the case here.  I'm leaning towards skip over fail, but I don't know what the right answer is.